### PR TITLE
Migration to React three fiber

### PR DESCRIPTION
**Pull request by Angatupyry**
## Migration to React three fiber

- Remove all react leaflet component 
- Using react three fiber instead. 

Pending:

- [ ] Labels overlapping. 
- [ ] Some hardcode z and scales values 
- [ ] Test from new react components
- [ ] It still happening that something the image doesn't load correctly and the frontend crash. I couldn't reproduce it properly yet.  
- [ ] Issue related #734 

[Screencast from 09-19-2023 07:41:42 PM.webm](https://github.com/open-rmf/rmf-web/assets/37310205/53f902aa-6ecf-41f0-8b78-0d57a912e21c)

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/react-components/lib/map/three-fiber/image-maker.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+export interface ImageThreeProps {
+  level: Level;
+  imageUrl: string;
+}
+
+export const ReactThreeFiberImageMaker = ({ level, imageUrl }: ImageThreeProps): JSX.Element => {
+  let texture: Texture | undefined = undefined;
+  try {
+    texture = useLoader(TextureLoader, imageUrl);
```
Comment (aaronchongth) [inline packages/react-components/lib/map/three-fiber/image-maker.tsx:14]: This is a tricky check failure (sorry I introduced this), however I'm unsure how we can be sure the `imageUrl` is accessible in the first place, these were some of the things I found, https://www.freecodecamp.org/news/how-to-validate-urls-in-javascript/.

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Thanks for helping with this huge migration! I think we're in pretty good shape now. There might be small issues here and there that pops up, but we can fix them later on.

I'll move this to a reviewable so CI can be run on it. Once we address all the check failures and CI passes, I think we are good.

File: packages/react-components/lib/map/three-fiber/image-maker.tsx (hunk 1 above)
Comment (Angatupyry) [reply packages/react-components/lib/map/three-fiber/image-maker.tsx]: It is currently working properly, could you try it?

File: packages/react-components/lib/map/three-fiber/shape-three-rendering.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+interface ShapeThreeRenderingProps {
+  position: [number, number, number];
+  color: string;
+  text?: string;
+  circleShape: boolean;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const debounce = <T extends (...args: any[]) => void>(
```
Comment (aaronchongth) [inline packages/react-components/lib/map/three-fiber/shape-three-rendering.tsx:12]: it looks like we're not using this function anywhere else at the moment, is it ok if we use
```
export const debounce = (callback: () => void, delay: number): (() => void) => {...}
```
This way we can avoid the eslint disabling

File: packages/dashboard/src/components/three-fiber/camera-control.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    // const handleWheel = (event: WheelEvent) => {
+    //   const delta = event.deltaY;
+    //   const sensitivity = 0.1;
+
+    //   setInternalZoom((prevZoom) => {
+    //     const newZoom = prevZoom - delta * sensitivity;
+    //     return Math.max(controls.minZoom, Math.min(controls.maxZoom, newZoom));
+    //   });
+
+    //   updateExternalStateDebounced(internalZoom);
```
Comment (aaronchongth) [inline packages/dashboard/src/components/three-fiber/camera-control.tsx:45]: I see that `debounce` is used here as well, I think when we implement it later on, we can just re-write it with a different callback instead of using a template. (if we still need it in the later implementation)

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Just a comment left regarding linting, otherwise LGTM

File: packages/dashboard/src/components/three-fiber/camera-control.tsx (hunk 3 above)
Comment (Angatupyry) [reply packages/dashboard/src/components/three-fiber/camera-control.tsx:45]: Yes, definitely.
Do you suggest removing the commented code?

File: packages/react-components/lib/map/three-fiber/shape-three-rendering.tsx (hunk 2 above)
Comment (Angatupyry) [reply packages/react-components/lib/map/three-fiber/shape-three-rendering.tsx]: Done 87b60ccd91cea03f780ba75f79d229b47a1edc8e

File: packages/dashboard/src/components/three-fiber/camera-control.tsx (hunk 3 above)
Comment (aaronchongth) [reply packages/dashboard/src/components/three-fiber/camera-control.tsx:45]: oh sorry, I didn't mention what the action item was. I think it is ok as it is for now, just commented to make sure we understand what to do with debouncing later on

File: packages/dashboard/src/components/map-app.tsx [hunk 4]
```diff
@@ -41,25 +42,31 @@ const debug = Debug('MapApp');
...
+      Waypoints: false,
+      Dispensers: false,
+      Ingestors: false,
+      Lifts: false,
+      Doors: false,
+      Trajectories: false,
+      Robots: false,
+      Labels: false,
+      PickupLabel: false,
+      DropoffLabel: false,
```
Comment (aaronchongth) [inline packages/dashboard/src/components/map-app.tsx:68]:
```suggestion
      Waypoint labels: false,
      Pickup point labels: false,
      Dropoff point labels: false,
```

My suggestion is not syntactically correct, but  the idea is to have spaces in the layer names.

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Thank you! Looks good, just a few tweaks and naming suggestions left

File: packages/dashboard/src/components/map-app.tsx (hunk 4 above)
Comment (Angatupyry) [reply packages/dashboard/src/components/map-app.tsx]: Done 5a39c12c876182611593353d365cb7a307ca6e89 !
I separate the labels in case we have more in the future. For now, I think they are just doors. 

Then the waypoint labels and the dropoff and pickup labels.

**Review by aaronchongth — APPROVED** [review]
I think we're in a good state! LGMT, thanks for this huge effort!
